src/uffd_handler/src/prefaulter.rs

```diff
@@ -21,11 +26,266 @@ impl UffdPrefaulter {
-                .ok_or_else(|| InvalidAmount(String::from("before"), before))?,
+                .checked_mul(KIB_MULT)
+                .ok_or_else(|| PrefaulterError::InvalidAmount(String::from("before"), before))?,
             bytes_after: after
-                .checked_add(KIB_MULT)
-                .ok_or_else(|| InvalidAmount(String::from("after"), before))?,
+                .checked_mul(KIB_MULT)
+                .ok_or_else(|| PrefaulterError::InvalidAmount(String::from("after"), after))?,
         })
     }
+
+    /// Return the start address of the chunk to prefault into memory.
+    /// When there is no customized prefaulting configured, the minimum chunk to copy to RAM
+    /// is the faulting page.
+    pub fn get_prefaulting_start_address(
+        &self,
+        page_fault_start_addr: usize,
+        region: &MemRegion,
+    ) -> Result<usize, HandlerError> {
+        let page_size = get_page_size().map_err(HandlerError::PageSize)?;
+
+        // Compute the start of the chunk to prefault according to the configuration
+        // and ensure it does not underflow.
+        let mut prefault_start_addr = page_fault_start_addr
+            .checked_sub(self.bytes_before)
+            .unwrap_or(region.mapping.base_host_virt_addr);
+        // Make sure that the start of the chunk to prefault is within the current region.
+        prefault_start_addr = max(prefault_start_addr, region.mapping.base_host_virt_addr);
+
+        // Page faults must be handled in units of pages. Because `bytes_before` is not constraint
+        // to be a multiple of page size, the start address must be rounded down to page boundary.
+        prefault_start_addr = page_start_of_addr(prefault_start_addr)?;
+
+        // If the start of the memory region has already been brought up into RAM
+        // (if for example a page fault was served by copying in RAM more than the faulting page)
+        // UFFD_COPY will return `EEXIST`. We need to find the first page not brought into memory
+        // after `prefault_start_addr`.
+        // This unwrap is safe because it is guaranteed that the key is the start of a page
+        // within the current region.
+        if *region.page_states.get(&prefault_start_addr).unwrap() == MemPageState::FromFile {
+            // Look for an uninitialized page up to the faulting page, which we know for
+            // sure that is not present in memory.
+            while prefault_start_addr < page_fault_start_addr {
+                prefault_start_addr += page_size;
+                // Found the first Uninitialized occurrence, so prefaulting will start from there.
+                // This unwrap is safe because it is guaranteed that the key is the start of a page
+                // within the current region.
+                if *region.page_states.get(&prefault_start_addr).unwrap()
+                    == MemPageState::Uninitialized
+                {
+                    break;
+                }
```

Comment:
Is there a possibility that the user, let's say, provides a `--before-kib 8` (which is 2 pages before) and those 2 pages before the actual page fault to be: the first one is `MemPageState::Uninitialized` but the following one is not `MemPageState::Uninitialized`? In this case, the `get_prefaulting_start_address()` will return the starting address of the `MemPageState::Uninitialized` page but the following one is not guaranteed to be also `MemPageState::Uninitialized`.  Would that be a concern thinking that we skipped the pages that are already in memory? Sorry, I'm not quite familiar with the code, yet.

Reply:
I _think_ this cant happen because any not-uninitialized page will have been brought in by the uffd handler, and the invariant of this uffd handler is that it always prefaults (e.g. if `--before-kb 8` is specified, then pages are always brought in in sequences of 3 pages - the page requested + 2 prefaulted ones). This means "islands" of initialized pages of size < 3 are impossible, and that's why the case you outline cannot happen. 

But I am also not sure. If this is the case, then that should definitely be mentioned in this method

Reply:
On a second thought, I need to think if my logic holds when we add the balloon device into the equation. I saw a test failing earlier and I think it might have something to do with the scenario outlined by @andreitraistaru above, coupled with the balloon device's work.

Reply:
Regarding the initial scenario (without the balloon device) it looks like, yeah...., it's not actually possible to reach the situation I mentioned if keeping the number of pages to prefault a constant (in case we'll allow changing `pages-before` at runtime, the problem will become possible but that's not the case yet, right?).

Regarding the second scenario (when adding the balloon device in discussion) I think it could be a problem depending on the placement of the pages taken from the balloon (I'm not really familiar with how the balloon works, but reading from docs, looks like the kernel is responsible with this so, I'm not sure how much control we have over it). I'm talking about a scenario like this:

Initially (3) was a faulting page some time ago, and we have a prefault of 2 pages before and 0 pages after ):

|.....|_FromFile_||_FromFile_|_FromFile_|_Uninitialized_|......|
               (1)                (2)               (3)                  (4)

After a memory allocation from guest, the balloon gives kernel another page; if the new page (5) is placed like this, it could be a problem:

|.....|_FromFile_||_FromFile_|_Uninitialized_|_FromFile_|_Uninitialized_|......|
               (1)                (2)               (5)                     (3)                  (4)

Now, if we get a page fault at (4), the prefaulter will return (5) as the start address to bring into memory which is not ok as long as we have (3) already in memory in between the start address and the actual page fault. But I'm still not convinced that such a situation would make sense or not.

Reply:
Yes, the scenario you described totally makes sense. We have no control over the pages that are claimed back by the host, so they could be placed anywhere really. 

Unfortunately, the balloon doesn't seem to work with UFFD for now. On balloon inflate, the removed pages are `madvise`'d with `MADV_DONTNEED`. When this happens, userfaultfd is supposed to receive also `Remove` event, apart from the `PageFault` events (see [here](https://man7.org/linux/man-pages/man2/ioctl_userfaultfd.2.html))
>If this feature is enabled, when the faulting process calls [madvise(2)](https://man7.org/linux/man-pages/man2/madvise.2.html) with the MADV_DONTNEED or MADV_REMOVE advice value to free a virtual memory area the userfaultfd monitor will receive an event of type UFFD_EVENT_REMOVE.

However, this doesn't happen, so uffd handler is unaware of the memory changes. The test that is failing our CI illustrates exactly this scenario: the handler fails because it tries to copy a page that has been claimed by the host, thus `UFFDIO_COPY` returns [`ENOENT`](https://man7.org/linux/man-pages/man2/ioctl_userfaultfd.2.html) and Firecracker freezes.

Reply:
I reduced the scope of this PR in order to unblock it, because it seems like we are going to need a different approach for ensuring that the faulting start is valid and that all pages in the computed range (at least up until the faulting page) are `Uninitialized`. My idea is to start checking from the faulting page down to the computed start and stop at the first occurrence of a non-`Uninitialized` page.

However, in order to unblock this PR, I removed the functionality of prefaulting pages before the faulting page. This way we always start from the faulting page and can only configure `pages-after`. Prefaulting pages before will be added in a subsequent PR. WDYT?

Reply:
Yes, I agree with the plan to split the work in multiple PRs and keep this PR for prefaulting pages after the faulting page only.